cli/lib/kontena/cli/stacks/yaml/validations.rb

```diff
@@ -31,7 +31,23 @@ def common_validations
         'external_links' => optional('array'),
         'mem_limit' => optional('string'),
         'mem_swaplimit' => optional('string'),
-        'environment' => optional(-> (value) { value.is_a?(Array) || value.is_a?(Hash) }),
+        'environment' => optional(-> (value) {
+          if value.is_a?(Hash)
+            value.all? do |k,v|
+              k.kind_of?(String) && (
+                v.kind_of?(String) ||
+                v.kind_of?(Integer) ||
+                v.kind_of?(TrueClass) ||
+                v.kind_of?(FalseClass) ||
+                v.nil?
+              )
+            end
+          elsif value.is_a?(Array)
+            value.all? { |v| v =~ /\A\S+(?<!\\)=.*/ }
```

Comment:
This should probably also validate that each array value is a `String`... now it might use the `=~` operator on a `Hash` or `Array`, which seems to return `nil` and thus fail validation, but that's odd.